bpf/lib/nat.h

```
@@ -1031,8 +1050,19 @@ snat_v4_rev_nat_handle_icmp_error(struct __ctx_buff *ctx,
 	if (!*state)
 		return NAT_PUNT_TO_STACK;
 
+	/*
+	 * The snat_v4_rewrite_headers() call only rewrites the checksum for
+	 * TCP and UDP.  UDP's checksum is covered by the RFC 792 inclusion
+	 * of the 1st 64 bits of the datagram following the IP header, but
+	 * TCP needs an additional 3 32-bit words to include the checksum.
+	 */
+	if (ctx_load_bytes(ctx, icmp_xlen_off, &icmp_xlen, sizeof(__u8)) < 0)
+		return DROP_INVALID;
+	icmp_has_full_l4_header = icmp_xlen >= ((tuple.nexthdr == IPPROTO_TCP) ? 3 : 0);
+
 	/* The embedded packet was SNATed on egress. Reverse it again: */
-	return snat_v4_rewrite_headers(ctx, tuple.nexthdr, (int)inner_l3_off, true, icmpoff,
+	return snat_v4_rewrite_headers(ctx, tuple.nexthdr, (int)inner_l3_off,
```

Wouldn't this mean that, for RFC 1191-compliant ICMP errors, if the original packet was TCP, the inner packet's port would not be modified? Wouldn't that be a problem?

---

Yes, that's correct - for a short ICMP error packet the inner L4 ports won't be modified for TCP.  For path MTU discovery the kernel is [looking](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ipv4/tcp_ipv4.c#n494) at the outer L3/L4 header, which allows the MTU [fixup call](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ipv4/tcp_ipv4.c#n591) to succeed.

---

Thanks for the clarification. The original datagram data includes the source and destination ports, so rewriting them is possible, but since we can't update the checksum, we can't guarantee integrity. So does that mean there's no point in rewriting them?

That said, it makes me wonder, what's the point of including such a incomplete data with no guarantee of integrity

---

My initial implementation of the fix left the call to snat_v4_rewrite_headers() intact with "true" for the inner L4 to go ahead and fix up the ports, but changed the check on its error return to ignore the error if there was not a complete header.  That way worked, too, and I can change it back to that if you'd rather see the port swap still happen.

---

@ysksuzuki @br4243 does this need more discussion to resolve the conversation? It's a bit unclear to me what is the next step on this PR.

---

@ysksuzuki Ping :slightly_smiling_face: